### PR TITLE
Update README

### PR DESCRIPTION
**Pull request by muneebahmed10**
*Description*:

Update README with build instructions for VS and QEMU

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,11 +1,6 @@
...
+[coreMQTT](https://github.com/FreeRTOS/coreMQTT) is an MIT licensed open source C MQTT client library for microcontrollers and small microprocessor based IoT devices. Its design is intentionally simple to ensure it has no dependency on any other library or operating system, and to better enable static analysis including [memory safety proofs](https://www.freertos.org/2020/02/ensuring-the-memory-safety-of-freertos-part-1.html). That simplicity and lack of operating system dependency (coreMQTT does not require multithreading at all) means coreMQTT does not build thread safety directly into its implementation. Instead thread safety must be provided by higher level software. This labs project implements a coreMQTT extension that provides that higher level functionality in the form of an MQTT agent (or MQTT daemon). While the implementation demonstrated here is currently specific to FreeRTOS, there are not many dependencies on FreeRTOS, meaning the implementation can easily be adapted for use with other operating systems.
```
Comment (aggarw13) [inline README.md:3]: Is this text relevant for this repository? Shouldn't it just say that this repository provides demo examples of using the coreMQTT Agent library to share a single MQTT connection amongst multiple application tasks? 
Maybe, after that a brief paragraph about the coreMQTT Agent library can be provided.

File: README.md [hunk 2]
```diff
@@ -14,18 +9,48 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+git clone git@github.com:FreeRTOS/coreMQTT-Agent-Demos.git --recurse-submodules
 ```
 
 If you have downloaded the repo without using the `--recurse-submodules` argument, you need to run:
 ```
 git submodule update --init --recursive
 ```
 
+## Getting started
+The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over the Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
```
Comment (aggarw13) [inline README.md:25]: OTA goes as **Over-the-Air** in the branding: https://docs.aws.amazon.com/freertos/latest/userguide/freertos-ota-dev.html
```suggestion
The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
```

File: README.md [hunk 3]
```diff
@@ -14,18 +9,48 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+
+**Eclipse IDE**
+
+1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.
+
+1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.
+
+1. Select `Project -> Build Project` from the Eclipse menu.
+
+In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
```
Comment (aggarw13) [inline README.md:52]:
```suggestion
In both cases of using **QEMU Cortex-M3** or **Eclipse IDE**, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
```

File: README.md [hunk 4]
```diff
@@ -14,18 +9,48 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+## Building Demos
+This repository contains build files for the FreeRTOS Windows port - able to be run with the free [Community version of Visual Studio](https://visualstudio.microsoft.com/vs/community/) - and an ARM Cortex M model in the [QEMU hardware emulator](https://www.qemu.org/).
+
+### Visual Studio
+
+1. From the Visual Studio IDE, open the `mqtt_multitask_demo.sln` Visual Studio solution file in the `build/VisualStudio/` directory.
+
+1. Select **Build Solution** from the IDE's **Build** menu
+
+### QEMU Cortex-M3
```
Comment (aggarw13) [inline README.md:36]: Will there be no instructions on how to run the executable in QEMU? It could be helpful for the reader to be pointed to a link for using QEMU

Comment (muneebahmed10) [reply README.md:36]: I would add it if I knew, but unfortunately I have not tried running it

File: README.md (hunk 1 above)
Comment (muneebahmed10) [reply README.md]: Added a few sentences before about this repository, but left this block here in a subsection

File: README.md [hunk 5]
```diff
@@ -1,6 +1,9 @@
...
+
+### What is coreMQTT?
+[coreMQTT](https://github.com/FreeRTOS/coreMQTT) is an MIT licensed open source C MQTT client library for microcontrollers and small microprocessor based IoT devices. Its design is intentionally simple to ensure it has no dependency on any other library or operating system, and to better enable static analysis including [memory safety proofs](https://www.freertos.org/2020/02/ensuring-the-memory-safety-of-freertos-part-1.html). That simplicity and lack of operating system dependency (coreMQTT does not require multithreading at all) means coreMQTT does not build thread safety directly into its implementation. Instead thread safety must be provided by higher level software. This project implements a coreMQTT extension that provides that higher level functionality in the form of an MQTT agent (or MQTT daemon). While the implementation demonstrated here is currently specific to FreeRTOS, there are not many dependencies on FreeRTOS, meaning the implementation can easily be adapted for use with other operating systems.
```
Comment (aggarw13) [inline README.md:6]: The last 2 sentences in this paragraph are not relevant to coreMQTT. 
```suggestion
[coreMQTT](https://github.com/FreeRTOS/coreMQTT) is an MIT licensed open source C MQTT client library for microcontrollers and small microprocessor based IoT devices. Its design is intentionally simple to ensure it has no dependency on any other library or operating system, and to better enable static analysis including [memory safety proofs](https://www.freertos.org/2020/02/ensuring-the-memory-safety-of-freertos-part-1.html). That simplicity and lack of operating system dependency (coreMQTT does not require multithreading at all) means coreMQTT does not build thread safety directly into its implementation. Instead thread safety must be provided by higher level software. The coreMQTT Agent library represents an extension to the coreMQTT library to provide the higher level thread-safe MQTT API functionality in the form of an MQTT agent (or MQTT daemon). ```

Comment (muneebahmed10) [reply README.md:6]: I'm going to edit your comment for `represents` -> `is` and since you needed a newline before "```"

File: README.md [hunk 6]
```diff
@@ -1,11 +1,9 @@
 ## MQTT Agent and Demo Tasks (including OTA) - Using coreMQTT
 
-Note:  The MQTT agent and associated example project are functional but not yet complete. Be aware the agent does not yet comply with our code quality standards and is not yet fully tested. There is a chance its APIs will change before its official first release.  Feedback, suggestions and comments are welcome on the [FreeRTOS support forum](https://forums.freertos.org).
+This repository provides demonstration examples of using the [coreMQTT Agent library](https://github.com/FreeRTOS/coreMQTT-Agent) with FreeRTOS. The coreMQTT Agent library is an extension on top of [coreMQTT](https://github.com/FreeRTOS/coreMQTT) that provides MQTT APIs with thread safety. The examples here share a single MQTT connection amongst multiple concurrent tasks, without requiring the management of any synchronization primitives from the application.
```
Comment (tanmoysen) [inline README.md:3]: I'd start with the most important points about the demo. For ex: "the demo uses OTA, Device Shadow, and Device Defender in multiple threads."

File: README.md [hunk 7]
```diff
@@ -1,11 +1,9 @@
 ## MQTT Agent and Demo Tasks (including OTA) - Using coreMQTT
```
Comment (tanmoysen) [inline README.md:1]: coreMQTT Agent?

File: README.md [hunk 8]
```diff
@@ -1,11 +1,9 @@
...
-
-## Getting started
-The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over the Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
+### What is coreMQTT?
```
Comment (tanmoysen) [inline README.md:5]: Why is this question relevant here? A user trying to use the demo would like to see info on coreMQTT-Agent rather than coreMQTT?

File: README.md [hunk 9]
```diff
@@ -14,18 +12,48 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.
+
+1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.
+
+1. Select `Project -> Build Project` from the Eclipse menu.
+
+In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
+
 ## Getting help
 You can use your Github login to get support from both the FreeRTOS community and directly from the primary FreeRTOS developers on our [active support forum](https://forums.freertos.org).  The [FAQ](https://www.freertos.org/FAQ.html) provides another support resource.
```
Comment (tanmoysen) [inline README.md:58]: "The FAQ provides another support resource" looks weird. Can we reword this to something like this? "You can find a list of frequently asked questions here."

File: README.md (hunk 6 above)
Comment (muneebahmed10) [reply README.md]: Reworded

File: README.md (hunk 7 above)
Comment (muneebahmed10) [reply README.md]: Reworded

File: README.md (hunk 8 above)
Comment (muneebahmed10) [reply README.md]: I will remove this paragraph as I think the above section has sufficient information on coreMQTT Agent's thread safety

File: README.md (hunk 9 above)
Comment (muneebahmed10) [reply README.md]: Reworded

File: README.md [hunk 10]
```diff
@@ -51,6 +54,22 @@ This repository contains build files for the FreeRTOS Windows port - able to be
 
 In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
 
+**Running the makefile build in QEMU**
+
+Use the QEMU command line contained in the [start_qemu.bat](https://github.com/FreeRTOS/coreMQTT-Agent-Demos/blob/main/build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated 'RTOSDemo.elf' file directly.  Alternatively use the command line contained in the [start_qemu_and_wait.bat](https://github.com/FreeRTOS/coreMQTT-Agent-Demos/blob/main/build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.
```
Comment (muneebahmed10) [inline README.md:59]: We should refer to it with a relative link since it's in the same repository. Also "QEMU command line" - can this just be "QEMU command"?
```suggestion
Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.
```

File: README.md [hunk 11]
```diff
@@ -51,6 +54,22 @@ This repository contains build files for the FreeRTOS Windows port - able to be
...
+
+Note these QEMU command lines assume there is a network tap driver called "TAP0".
+
+**Obtaing network access from QEMU on windows**
```
Comment (muneebahmed10) [inline README.md:63]:
```suggestion
**Obtaining network access from QEMU on Windows**
```

File: README.md [hunk 12]
```diff
@@ -51,6 +54,22 @@ This repository contains build files for the FreeRTOS Windows port - able to be
...
+
+1. Install the [OpenVPN TAP driver for Windows](https://openvpn.net/).
+
+1. Name the installed TAP interface "TAP0".  This can be done by selecting "Change adaptor settings" from the Windows' Network Properties window (Control Panel\Network and Internet\Network Connections) to view all network adaptors, selecting the installed TAP adapter, then pressing F2 to edit the adaptor's name.
```
Comment (muneebahmed10) [inline README.md:67]: There are two spellings of adapter here. I believe OpenVPN and Windows both use "e"
```suggestion
1. Name the installed TAP interface "TAP0".  This can be done by selecting "Change adapter settings" from Windows' Network Properties window (Control Panel\Network and Internet\Network Connections) to view all network adapters, selecting the installed TAP adapter, then pressing F2 to edit the adapter's name.
```

File: README.md [hunk 13]
```diff
@@ -51,6 +54,22 @@ This repository contains build files for the FreeRTOS Windows port - able to be
...
+
+1. From the same window, configure a real wired network interface to have a static IP address that is compatible with whichever IP address would otherwise be allocated by a DHCP server on your local network.
+
+1. Still in the same window, select both the real network adaptor and the TAP adapter at the same, then right click, and select "Bridge Connections" from the pop up menu.
```
Comment (muneebahmed10) [inline README.md:71]: Both adaptor and adapter here as well.
```suggestion
1. Still in the same window, select both the real network adapter and the TAP adapter at the same, then right click, and select "Bridge Connections" from the pop up menu.
```

File: README.md [hunk 14]
```diff
@@ -51,6 +54,22 @@ This repository contains build files for the FreeRTOS Windows port - able to be
...
+
+Note these QEMU command lines assume there is a network tap driver called "TAP0".
+
+**Obtaing network access from QEMU on windows**
+
+1. Install the [OpenVPN TAP driver for Windows](https://openvpn.net/).
+
+1. Name the installed TAP interface "TAP0".  This can be done by selecting "Change adaptor settings" from the Windows' Network Properties window (Control Panel\Network and Internet\Network Connections) to view all network adaptors, selecting the installed TAP adapter, then pressing F2 to edit the adaptor's name.
+
+1. From the same window, configure a real wired network interface to have a static IP address that is compatible with whichever IP address would otherwise be allocated by a DHCP server on your local network.
```
Comment (muneebahmed10) [inline README.md:69]: This is a little confusing. Does this mean get an address from the DHCP server and assign it as a static address, choose an available IP address from the DHCP server's range, or choose an IP address that does **not** conflict with the DHCP server's range? It sounds like it could be the second one, but I'm not sure
```suggestion
1. From the same window, configure a real wired network interface to have a static IP address that is within the range of IP addresses allocated by your local DHCP server.
```

File: README.md [hunk 15]
```diff
@@ -51,6 +54,22 @@ This repository contains build files for the FreeRTOS Windows port - able to be
 
 In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
 
+**Running the makefile build in QEMU**
+
+Use the QEMU command line contained in the [start_qemu.bat](https://github.com/FreeRTOS/coreMQTT-Agent-Demos/blob/main/build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated 'RTOSDemo.elf' file directly.  Alternatively use the command line contained in the [start_qemu_and_wait.bat](https://github.com/FreeRTOS/coreMQTT-Agent-Demos/blob/main/build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.
+
+Note these QEMU command lines assume there is a network tap driver called "TAP0".
```
Comment (muneebahmed10) [inline README.md:61]: Could they just be referred to as commands?
```suggestion
Note that these QEMU commands assume there is a network tap driver called "TAP0".
```

File: README.md [hunk 16]
```diff
@@ -14,20 +9,69 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
 If you have downloaded the repo without using the `--recurse-submodules` argument, you need to run:
 ```
 git submodule update --init --recursive
 ```
 
+## Getting started
+The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
+
+## Building Demos
+This repository contains both a Visual Studio project that uses the FreeRTOS Windows port and can be built using the free [Community version of Visual Studio](https://visualstudio.microsoft.com/vs/community/), and a GCC/makefile project that build the FreeRTOS ARM Cortex-M3 port and targets the [QEMU hardware emulator](https://www.qemu.org/).  The makefile can be built form the command line or the provided Eclipse project, and works on both Windows and Linux hosts.
```
Comment (muneebahmed10) [inline README.md:28]:
```suggestion
This repository contains both a Visual Studio project that uses the FreeRTOS Windows port and can be built using the free [Community version of Visual Studio](https://visualstudio.microsoft.com/vs/community/), and a GCC/makefile project that build the FreeRTOS ARM Cortex-M3 port and targets the [QEMU hardware emulator](https://www.qemu.org/).  The makefile can be built from the command line or the provided Eclipse project, and works on both Windows and Linux hosts.
```

File: README.md [hunk 17]
```diff
@@ -14,20 +9,69 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+
+1. Ensure that `arm-none-eabi-gcc` from the [GNU ARM Embedded Toolchain](https://developer.arm.com/tools-and-software/open-source-software/developer-tools/gnu-toolchain/gnu-rm/downloads) and GNU `make` are in your path.
+
+1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.
+
+1. Select `Project -> Build Project` from the Eclipse menu.
+
+In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
+
+**Running the makefile build in QEMU**
```
Comment (muneebahmed10) [inline README.md:57]: Might as well capitalize this here since this looks like a section title, and we also capitalize the Makefile file in the directory
```suggestion
**Running the Makefile build in QEMU**
```

File: README.md (hunk 10 above)
Comment (muneebahmed10) [reply README.md:59]: Should this be GBD or GDB? Also, what is the Eclipse debug configuration suitable for? Should it instead be "configuration suitable for debugging"?
```suggestion
Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GDB connection on port 1234.  The Eclipse project contains a configuration suitable for debugging.
```

File: README.md (hunk 4 above)
Comment (RichardBarry) [reply README.md:39]: Added.

File: README.md [hunk 18]
```diff
@@ -14,20 +9,69 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+
+1. Open the `build/Cortex-M3_MPS2_QEMU_GCC` directory as an Eclipse project.
+
+1. Select `Project -> Build Project` from the Eclipse menu.
+
+In both cases, the generated executable can be found at `build/Cortex-M3_MPS2_QEMU_GCC/output/RTOSDemo.elf`.
+
+**Running the Makefile build in QEMU**
+
+Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GBD connection on port 1234.  The Eclipse project contains a suitable debug configuration.
```
Comment (muneebahmed10) [inline README.md:59]:
```suggestion
Use the QEMU command contained in the [start_qemu.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu.bat) batch file to run the generated `RTOSDemo.elf` file directly.  Alternatively, use the command contained in the [start_qemu_and_wait.bat](build/Cortex-M3_MPS2_QEMU_GCC/start_qemu_and_wait.bat) batch file to start QEMU and wait for a GDB connection on port 1234.  The Eclipse project contains a configuration suitable for debugging.
```

File: README.md [hunk 19]
```diff
@@ -14,20 +9,69 @@ This repo uses [Git Submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodu
...
+git clone git@github.com:FreeRTOS/coreMQTT-Agent-Demos.git --recurse-submodules
 ```
 
 If you have downloaded the repo without using the `--recurse-submodules` argument, you need to run:
 ```
 git submodule update --init --recursive
 ```
 
+## Getting started
+The [documentation page](https://freertos.org/mqtt/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
```
Comment (muneebahmed10) [inline README.md:25]: I believe this is the new URL
```suggestion
The [documentation page](https://freertos.org/mqtt-agent/mqtt-agent-demo.html) for this repository contains information on the MQTT agent and the contained demo project.  There is also a [supplemental documentation page](https://freertos.org/ota/ota-mqtt-agent-demo.html) that describes how to run an Over-the-Air (OTA) update agent as one of the RTOS tasks that share the same MQTT connection.
```

**Review by RichardBarry — APPROVED** [review]

**Review by archigup — APPROVED** [review]
